lib/bolt_server/transport_app.rb

```diff
@@ -675,6 +681,68 @@ def make_winrm_target(target_hash)
       end
     end
 
+    # Returns the base64 encoded tar archive of plugin code that is needed to calculate
+    # custom facts
+    #
+    # @param versioned_project [String] the versioned_project to build the plugin tarball from
+    get '/project_facts_plugin_tarball' do
+      raise BoltServer::RequestError, "'versioned_project' is a required argument" if params['versioned_project'].nil?
+      content_type :json
+
+      # Inspired by Bolt::Applicator.build_plugin_tarball
+      start_time = Time.now
+
+      # Fetch the plugin files
+      plugin_files = in_bolt_project(params['versioned_project']) do |context|
+        files = {}
+
+        # Bolt also sets plugin_modulepath to user modulepath so do it here too for
+        # consistency
+        plugin_modulepath = context[:pal].user_modulepath
+        Puppet.lookup(:current_environment).override_with(modulepath: plugin_modulepath).modules.each do |mod|
+          search_dirs = []
+          search_dirs << mod.plugins if mod.plugins?
+          search_dirs << mod.pluginfacts if mod.pluginfacts?
+
+          files[mod] ||= []
+          Find.find(*search_dirs).each do |file|
+            files[mod] << file if File.file?(file)
+          end
+        end
+
+        files
+      end
+
+      # Pack the plugin files
+      sio = StringIO.new
+      begin
+        output = Minitar::Output.new(Zlib::GzipWriter.new(sio))
+
+        plugin_files.each do |mod, files|
+          parent = Pathname.new(mod.path).parent
```

Comment:
I have a hunch that there is a bug here somewhere around not using the module names (instead the paths) to build the archive structure. I have not fully formed this idea yet though... This looks intentional https://github.com/puppetlabs/bolt/blob/7b1caf6e314fabb21f84cdcf5900da5a2412869f/lib/bolt/applicator.rb#L309-L314 (using the name instead of the path)

Reply:
I think I missed this because my version of Bolt was out-of-date.. Thanks, updated.